### ISSUE TITLE
Trainer track_grad_norm always results in 0

### ISSUE THREAD

**Issue by kvhooreb** [williamFalcon/pytorch-lightning#272]
**Describe the bug**
The Trainer has a flag `track_grad_norm` which allows us to log the gradient norms to Tensorboard. This flag is checked in the `run_tng_epoch` function after the training step and validation step. However, the training step (`__run_tng_batch`) calls `model.optimizer_step()`, which, in the default implementation, calls `optimizer.zero_grad()`. This results in the tracked gradient norms to be always zero. 

Moreover, there is an optional `__run_evaluation` call in the validatin_step. This results in a call to `model.zero_grad()`, which I assume will also result in zero gradient norms. 

**To Reproduce**
1. Train a model with the `track_grad_norm` flag set to True and Tensorboard logging enabled
2. Go to Tensorboard
3. Check the gradient_norms

**Expected behavior**
The gradient norms should not always be zero.

**Desktop (please complete the following information):**
 - OS: Windows
 - Version 0.4.9

**Comment by kvhooreb** [comment]
I temporarily worked around this by overwriting the `optimizer_step()` function of my LightningModule to not set the gradients to zero, and moved that to the `on_batch_start` hook. However, it's clear that the validation runs also cause the gradients to be zero. See the image attached, where there is one validation batch every 5 training batches. The gradient clearly periodically drops to zero. 
![image](https://user-images.githubusercontent.com/7951058/65946306-29132c00-e436-11e9-8c91-2522c4df7212.png)

**Comment by williamFalcon** [comment]
good catch! 

1. evaluation does not calculate gradients, so it would be zero. the real solution is not to track grad norm during evaluation.
2. [ these lines](https://github.com/williamFalcon/pytorch-lightning/blob/480eed5cb61cb67037cac8b8eb6df9e0fe9286d6/pytorch_lightning/trainer/trainer.py#L1024-L1028), should be moved right before optimizer_step ([this line](https://github.com/williamFalcon/pytorch-lightning/blob/480eed5cb61cb67037cac8b8eb6df9e0fe9286d6/pytorch_lightning/trainer/trainer.py#L1230)).

Would love a PR to fix if you have some time!
